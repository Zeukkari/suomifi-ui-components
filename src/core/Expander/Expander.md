```jsx
import { Expander } from 'suomifi-ui-components';

<Expander title="Test expander" className="expander-test">
  Test expander
</Expander>;
```

```jsx
import { Expander } from 'suomifi-ui-components';

<ExpanderGroup OpenAll="Open all" CloseAll="Close all">
  <Expander title="Test expander 1">Test expander content 1</Expander>
  <Expander title="Test expander 2">Test expander content 2</Expander>
  <Expander title="Test expander 3">Test expander content 3</Expander>
</ExpanderGroup>;
```
