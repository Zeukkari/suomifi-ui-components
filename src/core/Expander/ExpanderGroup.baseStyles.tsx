import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../theme';
import { element, font, focus } from '../theme/reset';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme) => css`
  ${element({ theme })}
  display: flex;
  flex-direction: column;
  & > .fi-expander-group_panels {
    flex: none;

    & .fi-expander {
      margin-top: 0;
      margin-bottom: 0;
      transition: margin ${`${theme.transitions.basicTime}
        ${theme.transitions.basicTimingFunction}`};
      &.fi-expander--open {
        &:not(:first-of-type) {
          margin-top: ${theme.spacing.m};
        }
        &:not(:last-of-type) {
          margin-bottom: ${theme.spacing.m};
        }
      }
    }
  }

  & > .fi-expander-group_all-button {
    ${element({ theme })}
    ${font({ theme })('actionElementInnerTextBold')}
    ${focus({ theme })}
    flex: 1 1 auto;
    align-self: flex-end; 
    margin-left: auto;
    margin-bottom: ${theme.spacing.s};
    padding: ${theme.spacing.xs} 0;
    color: ${theme.colors.highlightBase};
    cursor: pointer;
  }
`,
);
