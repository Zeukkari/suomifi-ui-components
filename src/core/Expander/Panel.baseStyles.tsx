import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../theme';
import { element, font } from '../theme/reset';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme) => css`
  ${element({ theme })}
  ${font({ theme })('bodyText')}
  padding: ${theme.spacing.m};
  background-color: ${theme.colors.whiteBase};
`,
);
