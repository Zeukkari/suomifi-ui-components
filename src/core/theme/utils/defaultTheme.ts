import { suomifiTheme, ThemeProp } from '../';
import { AsPropType } from '../../../utils/typescript';

/**
 * Add default theme to props if not assinged,
 * also reassign styled as-prop to asProp (it would replace internal component usage with styled-tag)
 * @param props Component-props
 */
export const withDefaultTheme = <
  T extends {
    theme: ThemeProp;
    as?: AsPropType;
    asProp?: AsPropType;
    [k: string]: any;
  }
>({
  theme = suomifiTheme,
  as: asStyled,
  asProp,
  ...props
}: Partial<T>): T =>
  ({
    ...props,
    theme,
    ...(!!asProp || !!asStyled ? { asProp: asProp || asStyled } : {}),
  } as T);
