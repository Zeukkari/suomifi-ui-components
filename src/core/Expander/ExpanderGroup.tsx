import React from 'react';
import { default as styled } from 'styled-components';
import { Omit } from '../../utils/typescript';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp, InternalTokensProp } from '../theme';
import {
  ExpanderGroup as CompExpanderGroup,
  ExpanderGroupProps as CompExpanderGroupProps,
} from '../../components/Expander/ExpanderGroup';
import { Button, ButtonProps } from '../Button/Button';
import { baseStyles } from './ExpanderGroup.baseStyles';

const openAllButtonClassName = 'fi-expander-group_all-button';

type ButtonOrText = React.ReactElement<ButtonProps> | string;
interface OpenCloseAll {
  /** 'Open all'-component (Button) */
  OpenAll: ButtonOrText;
  /** 'Close all'-component (Button) */
  CloseAll: ButtonOrText;
}
export interface ExpanderGroupProps
  extends Omit<CompExpanderGroupProps, 'OpenAll' | 'CloseAll'>,
    TokensProp,
    OpenCloseAll {}

interface ExpanderOpenAllButtonProps extends ButtonProps, TokensProp {
  children: ButtonOrText;
}

const StyledExpanderGroup = styled(
  ({ tokens, ...passProps }: CompExpanderGroupProps & InternalTokensProp) => (
    <CompExpanderGroup {...passProps} />
  ),
)`
  ${props => baseStyles(props)};
`;

const OpenAllButton = ({
  children,
  ...passProps
}: ExpanderOpenAllButtonProps) => {
  const Component = React.Children.toArray(children)[0];
  if (typeof Component === 'string' || Component.type !== Button) {
    return (
      <Button.unstyled {...passProps} className={openAllButtonClassName}>
        {children}
      </Button.unstyled>
    );
  }
  return Component;
};

/**
 * <i class="semantics" />
 * Used for grouping expander panels
 */
export class ExpanderGroup extends React.Component<ExpanderGroupProps> {
  render() {
    const { OpenAll, CloseAll, ...passProps } = withSuomifiDefaultProps(
      this.props,
    );
    return (
      <StyledExpanderGroup
        {...passProps}
        OpenAll={
          <OpenAllButton tokens={passProps.tokens}>{OpenAll}</OpenAllButton>
        }
        CloseAll={
          <OpenAllButton tokens={passProps.tokens}>{CloseAll}</OpenAllButton>
        }
      />
    );
  }
}
