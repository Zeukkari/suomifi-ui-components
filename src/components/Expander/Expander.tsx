import React, { Component, ReactNode, Fragment } from 'react';
import { default as styled } from 'styled-components';
import { Button, ButtonProps } from '../Button/Button';
import { allStates } from '../../utils/css';
import { HtmlDiv, HtmlDivProps } from '../../reset';
import classnames from 'classnames';
import { ExpanderGroupConsumer, ExpanderProviderState } from './ExpanderGroup';

const baseClassName = 'fi-expander';
const openClassName = `${baseClassName}--open`;
const titleClassName = `${baseClassName}_title`;
const titleOpenClassName = `${titleClassName}--open`;
const titleNoTagClassName = `${titleClassName}--no-tag`;
const titleTagClassName = `${baseClassName}_title-tag`;
const contentBaseClassName = `${baseClassName}_content`;
const contentOpenClassName = `${contentBaseClassName}--open`;

const StyledPanel = styled((props: HtmlDivProps) => <HtmlDiv {...props} />)`
  display: block;
  width: 100%;
  max-width: 100%;
`;

interface CommonExpanderProps {
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Panel element content
   */
  children?: ReactNode;
}

interface StyledExpanderContentProps extends CommonExpanderProps {
  openState: boolean;
  hidden: boolean;
}

export const StyledExpanderContent = styled(
  ({ openState, className, ...passProps }: StyledExpanderContentProps) => (
    <StyledPanel
      {...passProps}
      className={classnames(className, baseClassName)}
    />
  ),
)`
  display: ${({ openState }) => (!!openState ? 'block' : 'none')};
`;

const StyledExpanderTitle = styled(
  ({ open, className, ...passProps }: ButtonProps & { open?: boolean }) => (
    <Button
      {...passProps}
      mouseNoFocus={true}
      className={classnames(className, { [titleOpenClassName]: !!open })}
    />
  ),
)`
  &,
  & * {
    cursor: pointer;
  }
  ${allStates('cursor: pointer;')}
`;

interface ExpanderState {
  openState: boolean;
}

export interface ExpanderProps extends CommonExpanderProps {
  /** Title for Panel */
  title: ReactNode;
  /** Title HTML-tag (h1, h2, h3 etc.)
   * @default none
   */
  titleTag?: string;
  open?: boolean;
  /** Properties for title-Button */
  titleProps?: ButtonProps & { open?: boolean };
  /** Properties for the content div */
  contentProps?: CommonExpanderProps;
  /** Default status of panel open
   * @default false
   */
  defaultOpen?: boolean;
  /** Event handler to execute when clicked */
  onClick?: ({ openState }: { openState: boolean }) => void;
  index?: number;
  expanderGroup?: boolean;
  consumer?: ExpanderProviderState;
}

const IfTitleTag = ({
  titleTag,
  children,
}: {
  titleTag?: string;
  children: ReactNode;
}) => (
  <Fragment>
    {!!titleTag
      ? React.createElement(titleTag, {
          children,
          className: titleTagClassName,
        })
      : children}
  </Fragment>
);

export class ExpanderItem extends Component<ExpanderProps> {
  state: ExpanderState = {
    openState:
      this.props.defaultOpen !== undefined ? this.props.defaultOpen : false,
  };

  handleClick = () => {
    const {
      open,
      onClick,
      consumer: { onClick: consumerOnClick } = { onClick: undefined },
      index,
      expanderGroup,
    } = this.props;
    if (!!expanderGroup && !!consumerOnClick && index !== undefined) {
      consumerOnClick(index);
    } else {
      const openState = !this.state.openState;
      const notControlled = open === undefined;
      if (notControlled) {
        this.setState({ openState });
      }
      if (!!onClick) {
        onClick({ openState: notControlled ? openState : !!open });
      }
    }
  };

  render() {
    const {
      open,
      defaultOpen,
      onClick,
      className,
      children,
      title,
      titleTag,
      titleProps,
      index,
      expanderGroup: dissmissExpanderGroup,
      consumer: { openPanels } = { openPanels: undefined },
      contentProps: { className: contentClassName, ...contentPassProps } = {
        className: undefined,
      },
      ...passProps
    } = this.props;
    const localOpenState = () =>
      open !== undefined ? !!open : this.state.openState;
    const openState =
      index !== undefined && !!openPanels
        ? openPanels.includes(index)
        : localOpenState();

    return (
      <StyledPanel
        {...passProps}
        className={classnames(className, baseClassName, {
          [openClassName]: !!openState,
        })}
      >
        <IfTitleTag titleTag={titleTag}>
          <StyledExpanderTitle
            onClick={this.handleClick}
            className={classnames(titleClassName, {
              [titleNoTagClassName]: !titleTag,
            })}
            aria-expanded={!!openState}
            {...titleProps}
            open={openState}
          >
            {title}
          </StyledExpanderTitle>
        </IfTitleTag>
        <StyledExpanderContent
          {...contentPassProps}
          openState={openState}
          className={classnames(contentBaseClassName, contentClassName, {
            [contentOpenClassName]: !!openState,
          })}
          hidden={!openState}
          key={String(openState)}
          aria-hidden={!openState}
        >
          {children}
        </StyledExpanderContent>
      </StyledPanel>
    );
  }
}

export class Expander extends Component<ExpanderProps> {
  render() {
    return !!this.props.expanderGroup ? (
      <ExpanderGroupConsumer>
        {ExpanderGroupConsumer => (
          <ExpanderItem {...this.props} consumer={ExpanderGroupConsumer} />
        )}
      </ExpanderGroupConsumer>
    ) : (
      <ExpanderItem {...this.props} />
    );
  }
}
