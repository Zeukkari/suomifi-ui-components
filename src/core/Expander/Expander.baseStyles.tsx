import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../theme';
import { button } from '../theme/reset';
import { absolute } from '../../utils/css';
import { padding } from '../theme/utils';
import { ExpanderProps } from './Expander';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme & Partial<ExpanderProps>) => {
    return css`
  ${absolute('before')}
  position: relative;
  padding: 0;
  border-radius: ${theme.radius.basic};
  box-shadow: ${theme.shadows.panelShadow};

  &:before {
    background-color: ${theme.colors.highlightLight53};
    opacity: 0;
  }

  &.fi-expander--open:before {
    opacity: 1;
    transition: opacity ${`${theme.transitions.basicTime}
      ${theme.transitions.basicTimingFunction}`};
  }

  & .fi-expander_title {
    ${button({ theme })}
    position: relative;
    display: block;
    width: 100%;
    &--no-tag {
      ${padding({ theme })('m', 'xl', 'm', 'm')}
      color: ${theme.colors.highlightBase};
    }
  }
  & .fi-expander_title-icon {
    position: absolute;
    top: 0;
    right: 0;
    margin: ${theme.spacing.m};
  }
  & .fi-expander_title--open .fi-expander_title-icon,
  & .fi-expander_title-icon--open {
    transform: rotate(-180deg);
  }

  & > .fi-expander_content {
    position: relative;
    display: block;
    height: 0;
    overflow: hidden;
    transform: scaleY(0);
    transform-origin: top;
    transition: all ${`${theme.transitions.basicTime}
      ${theme.transitions.basicTimingFunction}`};
    will-change: transition, height;
    &:not(.fi-expander_content--no-padding) {
      padding: 0 ${theme.spacing.m};
    }
    &.fi-expander_content--open {
      height: 10%;
      overflow: visible;
      /* This is very robust - cannot animate dynamic height with height-definition */
      animation: fi-expander_content-anim ${theme.transitions.basicTime}
        ${theme.transitions.basicTimingFunction} 1 forwards;
      &:not(.fi-expander_content--no-padding) {
        ${padding({ theme })('0', 'm', 'm', 'm')}
      }
    }
    @keyframes fi-expander_content-anim {
      0% {
        height: auto;
        transform: scaleY(0);
      }
      100% {
        transform: scaleY(1);
      }
    }
  }
`;
  },
);
