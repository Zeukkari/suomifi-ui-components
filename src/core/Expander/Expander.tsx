import React, { Component, Fragment } from 'react';
import { default as styled } from 'styled-components';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp, InternalTokensProp } from '../theme';
import { baseStyles as panelBaseStyles } from './Panel.baseStyles';
import { baseStyles } from './Expander.baseStyles';
import {
  Expander as CompExpander,
  ExpanderProps as CompExpanderProps,
} from '../../components/Expander/Expander';
import { Icon } from '../Icon/Icon';
import classnames from 'classnames';
import { ExpanderGroupProps, ExpanderGroup } from './ExpanderGroup';

const iconClassName = 'fi-expander_title-icon';
const iconOpenClassName = `${iconClassName}--open`;
const noPaddingClassName = `fi-expander_content--no-padding`;

type ExpanderVariant = 'expander' | 'expanderGroup';

export interface ExpanderProps extends CompExpanderProps, TokensProp {
  /** Remove padding from expandable content area (for background usage with padding in given container etc.) */
  noPadding?: boolean;
  /**
   * 'expander' | 'expanderGroup'
   * @default expander
   */
  variant?: ExpanderVariant;
}

const StyledExpander = styled(
  ({ tokens, noPadding, ...passProps }: ExpanderProps & InternalTokensProp) => {
    return (
      <CompExpander
        {...passProps}
        contentProps={{
          className: classnames({ [noPaddingClassName]: noPadding }),
        }}
      />
    );
  },
)`
  ${props => panelBaseStyles(props)};
  ${props => baseStyles(props)};
`;

interface ExpanderState {
  openState: boolean;
}

/**
 * <i class="semantics" />
 * Used for openable panel
 */
export class Expander extends Component<ExpanderProps> {
  static group = (props: ExpanderGroupProps) => {
    return <ExpanderGroup {...withSuomifiDefaultProps(props)} />;
  };

  /** State is only used to update the caret-icon */
  state: ExpanderState = {
    openState:
      this.props.defaultOpen !== undefined ? this.props.defaultOpen : false,
  };

  handleClick = () => {
    const { open, onClick } = this.props;
    const { openState } = this.state;
    const notControlled = open === undefined;
    if (notControlled) {
      this.setState({ openState: !openState });
    }
    if (!!onClick) {
      onClick({ openState: notControlled ? !openState : !!open });
    }
  };

  render() {
    const {
      variant,
      open,
      title,
      titleTag,
      ...passProps
    } = withSuomifiDefaultProps(this.props);
    const notControlled = open === undefined;
    const openState = !notControlled ? open : this.state.openState;

    if (variant === 'expander' && 'title' in passProps) {
      return <Expander {...passProps as ExpanderProps} />;
    }

    return (
      <StyledExpander
        {...passProps}
        onClick={this.handleClick}
        open={open}
        titleTag={titleTag}
        title={
          <Fragment>
            {title}
            {!titleTag && (
              <Icon
                icon="chevronDown"
                className={classnames(iconClassName, {
                  [iconOpenClassName]: openState,
                })}
                color={passProps.tokens.colors.highlightBase}
              />
            )}
          </Fragment>
        }
      />
    );
  }
}
