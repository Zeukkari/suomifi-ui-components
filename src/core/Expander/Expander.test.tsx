import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../utils/test/axe';

import { Expander } from './Expander';
import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Panel.baseStyles';

const TestExpander = (
  <Expander
    title="Test expander"
    titleProps={{ 'data-testid': 'expander-title' }}
    className="expander-test"
  >
    Test expander content
  </Expander>
);

test('calling render with the same component on the same container does not remount', () => {
  const panelRenderer = render(TestExpander);
  const { getByTestId, container, rerender } = panelRenderer;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('expander-title').textContent).toBe('Test expander');

  // re-render the same component with different props
  rerender(
    <Expander
      title="Test expander two"
      titleProps={{ 'data-testid': 'expander-title-2' }}
    >
      Test expander content
    </Expander>,
  );
  expect(getByTestId('expander-title-2').textContent).toBe('Test expander two');
});

test('CSS export', () => {
  const css = cssFromBaseStyles(baseStyles);
  expect(css).toEqual(expect.stringContaining('background-color'));
});

test('should not have basic accessibility issues', axeTest(TestExpander));
