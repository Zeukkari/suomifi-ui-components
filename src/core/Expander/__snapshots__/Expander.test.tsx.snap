// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c5:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c5::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c5::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c5::-webkit-inner-spin-button {
  height: auto;
}

.c5::-webkit-outer-spin-button {
  height: auto;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4.fi-button--disabled {
  cursor: not-allowed;
}

.c1 {
  display: block;
  width: 100%;
  max-width: 100%;
}

.c7 {
  display: none;
}

.c3,
.c3 * {
  cursor: pointer;
}

.c3:hover,
.c3:active,
.c3:focus,
.c3:focus-within {
  cursor: pointer;
}

.c6 {
  display: inline-block;
  vertical-align: baseline;
}

.c0 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  padding: 16px;
  background-color: hsl(0,0%,100%);
  position: relative;
  position: relative;
  padding: 0;
  border-radius: 2px;
  box-shadow: 0 1px 2px 0 rgba(41,41,41,0.14),0 1px 5px 0 rgba(41,41,41,0.12);
}

.c0:before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.c0:before {
  background-color: hsl(212,63%,98%);
  opacity: 0;
}

.c0.fi-expander--open:before {
  opacity: 1;
  -webkit-transition: opacity 50ms cubic-bezier(0.28,0.84,0.42,1);
  transition: opacity 50ms cubic-bezier(0.28,0.84,0.42,1);
}

.c0 .fi-expander_title {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  line-height: 1;
  position: relative;
  display: block;
  width: 100%;
}

.c0 .fi-expander_title:focus {
  outline: 0;
  position: relative;
}

.c0 .fi-expander_title:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid hsl(23,82%,53%);
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 hsl(23,82%,53%);
  z-index: 9999;
}

.c0 .fi-expander_title:focus:not(:focus-visible):after {
  content: none;
}

.c0 .fi-expander_title--no-tag {
  padding-top: 16px;
  padding-right: 64px;
  padding-bottom: 16px;
  padding-left: 16px;
  color: hsl(212,63%,45%);
}

.c0 .fi-expander_title-icon {
  position: absolute;
  top: 0;
  right: 0;
  margin: 16px;
}

.c0 .fi-expander_title--open .fi-expander_title-icon,
.c0 .fi-expander_title-icon--open {
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
}

.c0 > .fi-expander_content {
  position: relative;
  display: block;
  height: 0;
  overflow: hidden;
  -webkit-transform: scaleY(0);
  -ms-transform: scaleY(0);
  transform: scaleY(0);
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  will-change: transition,height;
}

.c0 > .fi-expander_content:not(.fi-expander_content--no-padding) {
  padding: 0 16px;
}

.c0 > .fi-expander_content.fi-expander_content--open {
  height: 10%;
  overflow: visible;
  -webkit-animation: fi-expander_content-anim 50ms cubic-bezier(0.28,0.84,0.42,1) 1 forwards;
  animation: fi-expander_content-anim 50ms cubic-bezier(0.28,0.84,0.42,1) 1 forwards;
}

.c0 > .fi-expander_content.fi-expander_content--open:not(.fi-expander_content--no-padding) {
  padding-top: 0;
  padding-right: 16px;
  padding-bottom: 16px;
  padding-left: 16px;
}

<div
  class="expander-test c0 fi-expander c1 c2"
>
  <button
    aria-disabled="false"
    aria-expanded="false"
    class="fi-button fi-expander_title fi-expander_title--no-tag c3 c4 c5"
    data-testid="expander-title"
    tabindex="0"
    type="button"
  >
    Test expander
    <svg
      aria-hidden="true"
      class="fi-expander_title-icon c6"
      fill="hsl(212, 63%, 45%)"
      focusable="false"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M13 15.6c-.3.3-.6.4-1 .4s-.7-.1-1-.4l-5.6-5.3c-.5-.5-.5-1.4 0-1.9s1.4-.5 2 0l4.6 4.4 4.6-4.4c.5-.5 1.4-.5 2 0 .5.5.5 1.4 0 1.9L13 15.6z"
      />
    </svg>
  </button>
  <div
    aria-hidden="true"
    class="fi-expander_content c7 fi-expander c1 c2"
    hidden=""
  >
    Test expander content
  </div>
</div>
`;
